presto-bytecode/src/main/java/com/facebook/presto/bytecode/ClassDefinition.java

```diff
@@ -122,6 +123,11 @@ public List<MethodDefinition> getMethods()
         return ImmutableList.copyOf(methods);
     }
 
+    public int getAndIncrementMethodCounter()
```

Comment:
It's very weird to have a method counter here which doesn't apply to all methods in the ClassDefinition. It's probably better to either say whether you can add a counter to the caller of `generateMethodsForLambda` on the `methodNamePrefix`, or just add a static counter in the method to always increment it no matter who calls.

Reply:
I added an unit test, but the unit test does not fail for some reasons. I am very unfamiliar with this code, but block on the release as this fails few queries.

Reply:
The query in the issue doesn't fail without this fix?

Reply:
The query fails on verifier cluster consistently, but the query does not fail without the fix on the master. There is some combination of session properties required and not sure how to configure them in the TestSqlFunctions as I am very unfamiliar with this code. 

This is the failure in the verifier cluster.

```
Caused by: java.lang.IllegalArgumentException: Method with same name and signature already exists: sqllambda_1
	at com.facebook.presto.bytecode.ClassDefinition.declareMethod(ClassDefinition.java:284)
	at com.facebook.presto.sql.gen.LambdaBytecodeGenerator.defineLambdaMethod(LambdaBytecodeGenerator.java:251)
	at com.facebook.presto.sql.gen.LambdaBytecodeGenerator.preGenerateLambdaExpression(LambdaBytecodeGenerator.java:234)
	at com.facebook.presto.sql.gen.LambdaBytecodeGenerator.generateMethodsForLambda(LambdaBytecodeGenerator.java:181)
	at com.facebook.presto.sql.gen.LambdaBytecodeGenerator.generateMethodsForLambda(LambdaBytecodeGenerator.java:128)
	at com.facebook.presto.sql.gen.RowExpressionCompiler$Visitor.visitCall(RowExpressionCompiler.java:148)
	at com.facebook.presto.sql.gen.RowExpressionCompiler$Visitor.visitCall(RowExpressionCompiler.java:108)
	at com.facebook.presto.spi.relation.CallExpression.accept(CallExpression.java:125)
	at com.facebook.presto.sql.gen.RowExpressionCompiler.compile(RowExpressionCompiler.java:105)
	at com.facebook.presto.sql.gen.RowExpressionCompiler.compile(RowExpressionCompiler.java:98)
	at com.facebook.presto.sql.gen.LambdaBytecodeGenerator.defineLambdaMethod(LambdaBytecodeGenerator.java:255)
	at com.facebook.presto.sql.gen.LambdaBytecodeGenerator.preGenerateLambdaExpression(LambdaBytecodeGenerator.java:234)
	at com.facebook.presto.sql.gen.LambdaBytecodeGenerator.generateMethodsForLambda(LambdaBytecodeGenerator.java:181)
	at com.facebook.presto.sql.gen.LambdaBytecodeGenerator.generateMethodsForLambda(LambdaBytecodeGenerator.java:128)
	at com.facebook.presto.sql.gen.RowExpressionCompiler$Visitor.visitCall(RowExpressionCompiler.java:148)
	at com.facebook.presto.sql.gen.RowExpressionCompiler$Visitor.visitCall(RowExpressionCompiler.java:108)
	at com.facebook.presto.spi.relation.CallExpression.accept(CallExpression.java:125)
	at com.facebook.presto.sql.gen.RowExpressionCompiler.compile(RowExpressionCompiler.java:105)
	at com.facebook.presto.sql.gen.RowExpressionCompiler.compile(RowExpressionCompiler.java:98)
	at com.facebook.presto.sql.gen.CursorProcessorCompiler.generateProjectMethod(CursorProcessorCompiler.java:330)
	at com.facebook.presto.sql.gen.CursorProcessorCompiler.generateMethods(CursorProcessorCompiler.java:146)
	at com.facebook.presto.sql.gen.ExpressionCompiler.compileProcessor(ExpressionCompiler.java:231)
	at com.facebook.presto.sql.gen.ExpressionCompiler.compile(ExpressionCompiler.java:210)
	at com.facebook.presto.sql.gen.ExpressionCompiler.lambda$new$0(ExpressionCompiler.java:77)
	at com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:165)
	at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3528)
```

Reply:
@kaikalur what is failing under what condition? This cannot be a recently introduced new bug right (if it is we can just revert the change?) so why is this blocking release?

Reply:
We had our internal config forcing inline_sql_functions to true and we want to get rid of that. But when its false the query as mentioned in the issue fails.

Reply:
So this is not a release blocker right? But don't we always want inline_sql_functions to be true anyways? The default for inline_sql_functions is true.

Reply:
@arunthirupathi Looks like this is pagesourcecompiler specific so the array has to come from a table - values, literals etc don't show the issue. Also did you try adding two functions like have in the issue?

Reply:
I see, I was using CTE and using values. Let me try with the tables.

```
WITH tmp AS (
    SELECT
        1 AS id,
        ARRAY[CAST(row(ARRAY[1, 2]) AS ROW(col1 ARRAY(int)))] AS hist
    UNION ALL
    SELECT
        2 AS id,
        ARRAY[CAST(row(ARRAY[3, 4]) AS ROW(col1 ARRAY(int)))] AS hist
)
SELECT 
    id, 
    "array_sum"(
      "transform"(
        hist, 
        (entity) -> "array_sum"(
          "transform"(
            entity.col1, 
            (check) -> check * 2
          )
        )
      )
    ) checks
FROM tmp
```

Reply:
I am able to repro this on the verifier cluster consistently. But the unit test eludes me. 

I am able to see the bug in the code. The code tries to use a counter from the map which is extensively copied. 

@rongrong  do you have concerns merging this fix and unblocking the release ?

Reply:
But were you able to verify on the verifier with your fix no crash (with or without inline_sql_functio?)

Reply:
Yes, I was able to verify it on the verifier cluster after the fix, it does not crash and returns the results correctly.

Reply:
Problem with reproing on the unit test is multi fold.

1. I don't know how to set up the session property in the TestSqlFunctions.
2. Creating table and running the query on the table.

Reply:
> Problem with reproing on the unit test is multi fold.
> 
> 1. I don't know how to set up the session property in the TestSqlFunctions.
> 2. Creating table and running the query on the table.

You can try to add the test in: TestHiveIntegrationSmokeTest that lets you create the table. Also look at AbstractTestqueries.java where we create session objects with specific session param settings and run queries using that.